File: RA 1.2/Actividad 3.py
class Producto: 
    def __init__(self, codigo, nombre, precio, categoria):
        self.codigo = codigo
        self.nombre = nombre
        self.precio = precio
        self.categoria = categoria

    def __str__(self):
        return f"{self.codigo} | {self.nombre} | {self.precio:.2f} | {self.categoria}"
    
class Inventario: 
    def __init__(self):
        self.productos = []
        self.estadisticas = [
        ["Electronica", 0, 0.0],
        ["Ropa", 0, 0.0],
        ["Alimentos", 0, 0.0],
        ["Hogar", 0, 0.0]
        ]

    def agregar_producto(self, producto): 
        self.productos.append(producto)
        self._actualizar_estadisticas(producto)
        print(f"Producto '{producto.nombre}' agregado correctamente")

    def _actualizar_estadisticas(self, producto):
        for i in range(len(self.estadisticas)):
            if self.estadisticas[i][0] == producto.categoria: 
                self.estadisticas[i][1] += 1
                self.estadisticas[i][2] += producto.precio
                break

    def buscar_por_categoria(self, categoria): 
        resultados= []
        for producto in self.productos:
            if producto.categoria.lower() == categoria.lower():
                resultados.append(producto)
        return resultados 
    
    def mostrar_inventario(self): 
        if not self.productos: 
            print("Inventario vacio")
            return
        
        print("\n" + "="*60)
        print("LISTA DE PRODUCTOS (Arreglo Unidimensional)")
        print("="*60)
        print("Codigo | Nombre | Precio | Categoria")
        print("-"*60)
        for producto in self.productos:
            print(producto)

    def mostrar_estadisticas(self): 
        print("\n" + "="*60)
        print("ESTADISTICAS POR CATEGORIA (Matriz - Arreglo Bidimensional)")
        print("="*60)
        print("Categoria | Cantidad | Valor Total")
        print("-"*60)
        for fila in self.estadisticas:
            print(f"{fila[0]:11} | {fila[1]:8} | ${fila[2]:.2f}")

##################################################################################
inventario= Inventario()

productos_iniciales = [
Producto("001", "Laptop", 12000.00, "Electronica"),
Producto("002", "Camiseta", 250.00, "Ropa"),
Producto("003", "Arroz", 85.50, "Alimentos"),
Producto("004", "Sarten", 350.00, "Hogar"),
Producto("005", "Smartphone", 8000.00, "Electronica"),
Producto("006", "Pantalon", 450.00, "Ropa")
]

for prod in productos_iniciales: 
    inventario.agregar_producto(prod)

inventario.mostrar_inventario()

print("\n" + "="*60)
categoria_buscar= input("Buscar productos por categoria: ")
resultados= inventario.buscar_por_categoria(categoria_buscar)

if resultados: 
    print(f"\nProductos en categoria '{categoria_buscar}':")
    
    for prod in resultados:
        print(f" - {prod.nombre} - ${prod.precio:.2f}")
else:
    print(f"No se encontraron productos en la categoria '{categoria_buscar}'")
    inventario.mostrar_estadisticas()

File: RA 1.2/Actividad 3.cpp
#include <iostream> 
#include <vector>
#include <map> 
#include <algorithm>
using namespace std; 

class Producto {
    public: 
        string codigo; 
        string nombre; 
        double precio; 
        string categoria; 

    public: 
        Producto(string c, string n, double p, string ca): codigo(c), nombre(n), precio(p), categoria(ca) {}
        ~Producto() {}

        friend ostream& operator<<(ostream& os, const Producto& p) {
            os <<p.codigo<<" | "<<p.nombre<<" | $"<<p.precio<<" | "<<p.categoria; 
            return os;
        }
};

class Inventario { 
    protected: 
        vector<Producto> productos;
        struct Estadisticas {
            string categoria;
            int cantidad;
            double precio;
        };
        vector<Estadisticas> estadisticas= {
            {"Electronica", 0, 0.0},
            {"Ropa", 0, 0.0},
            {"Alimentos", 0, 0.0},
            {"Hogar", 0, 0.0}
        };

    public: 
        void agregar_producto(Producto producto) {
            productos.push_back(producto); 
            this->actualizar_estadisticas(producto); 
            cout<<"Producto '"<<producto.nombre<<"' agregado correctamente"<<endl;
        }

        void actualizar_estadisticas(Producto producto) {
            bool encontrado;
            for(size_t i= 0; i < estadisticas.size(); ++i) {
                if(estadisticas[i].categoria == producto.categoria) {
                    estadisticas[i].cantidad += 1; 
                    estadisticas[i].precio += producto.precio;
                    encontrado= true;
                    break;
                }
            }
            if(!encontrado) {
                estadisticas.push_back({producto.categoria, 1, producto.precio});
            }
        }

        vector<Producto> buscar_por_categoria(string categoria) {
            vector<Producto> resultados; 
            for(auto producto : this->productos) {
                if(producto.categoria == categoria) {
                    resultados.push_back(producto);
                }
            }
            return resultados;
        }

        void mostrar_inventario() {
            if (this->productos.empty()) {
                cout<<"Inventario vacio"<<endl; 
                return; 
            }
            cout<<"\n"<<string(60, '=')<<endl; 
            cout<<"LISTA DE PRODUCTOS (Arreglo (vector en C++) Unidimensional)"<<endl; 
            cout<<string(60, '=')<<endl; 
            cout<<"Codigo | Nombre | Precio | Categoria"<<endl; 
            cout<<string(60, '-')<<endl;
            for(auto producto : this->productos) {
                cout<<producto<<endl;
            }
        }

        void mostrar_estadisticas() {
            cout<<"\n"<<string(60, '=')<<endl; 
            cout<<"ESTADISTICAS POR CATEGORIA (Matriz - Arreglo Bidimensional )"<<endl; 
            cout<<string(60, '=')<<endl; 
            cout<<"Categoria | Cantidad | Valor Total"<<endl; 
            cout<<string(60, '-')<<endl;
            for(size_t i= 0; i < estadisticas.size(); ++i) {
                cout<<estadisticas[i].categoria<<" | "<<estadisticas[i].cantidad<<" | $"<<estadisticas[i].precio<<endl;
            }
        }
};

//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
int main() {
    Inventario inventario; 

    vector<Producto> productos_iniciales= {
        Producto("001", "Laptop", 12000.00, "Electronica"),
        Producto("002", "Camiseta", 250.00, "Ropa"),
        Producto("003", "Arroz", 85.50, "Alimentos"),
        Producto("004", "Sarten", 350.00, "Hogar"),
        Producto("005", "Smartphone", 8000.00, "Electronica"),
        Producto("006", "Pantalon", 450.00, "Ropa"),
        Producto("007", "Albedo body pillow", 750.00, "Anime")
    }; 

    for(auto prod : productos_iniciales) {
        inventario.agregar_producto(prod);
    }

    inventario.mostrar_inventario();
    inventario.mostrar_estadisticas();

    cout<<string(60, '=')<<endl;
    string categoria_buscar; 
    cout<<"Buscar productos por categoria: ";
    getline(cin>> ws, categoria_buscar); 
    vector<Producto> resultados= inventario.buscar_por_categoria(categoria_buscar);

    if(!resultados.empty()) {
        cout<<"\nProductos en categoria "<<categoria_buscar<<": "<<endl;
        for(auto prod : resultados) {
            cout<<" - "<<prod.nombre<<" - $"<<prod.precio<<endl;
        }
    } else { 
        cout<<"No se encontraron productos en la categoria '"<<categoria_buscar<<"' "<<endl;
        inventario.mostrar_estadisticas();
    }
    return 0;
}